util/system/path.go

```diff
@@ -220,5 +222,23 @@ func CheckSystemDriveAndRemoveDriveLetter(path string, inputOS string) (string,
 	//
 	// We must return the second element of the split path, as is, without attempting to convert
 	// it to an absolute path. We have no knowledge of the CWD; that is treated elsewhere.
-	return ToSlash(filepath.Clean(parts[1]), inputOS), nil
+	cleanedPath := filepath.Clean(parts[1])
+	return ToSlash(keepTrailingSlash(cleanedPath, parts[1], inputOS), inputOS), nil
+}
+
+// Checks if a path other than root has a trailing slash
+// and keeps it. See https://github.com/moby/buildkit/issues/5249
+// Expects cleanedPath to have gone through filepath.Clean()
+// which returns the path with platform specific separators.
+func keepTrailingSlash(cleanedPath, origPath, inputOS string) string {
```

Comment:
Nit: Would it make sense to create a `CleanPath()` function in this package that incorporates `keepTrailingSlash()`? We could then replace calls to `filepath.Clean()` with `CleanPath()`

Perhaps something like: https://go.dev/play/p/lcvDPMnT8xS ? What do you think?

Reply:
I like that, looks neat, let me do that. Thanks!

Reply:
Perhaps the only modification I'll do is to make our `CleanPath` take an additional `keepSlash` bool param, to be consistent with `NormalizePath`; as much as rn, it's always `true`.

Reply:
So, I've given it a little of some thought, I think we can minimize the surface area to just focus on our adaptation of `filepath.Clean` to retain the trailing slash. If you don't need to retain the trailing slash, then you just call `filepath.Clean`. I think this will improve composability and avoid any future surprises.

I have also dropped the `inputOS` param since the initial intention for it was for the `ToSlash` function. So we clean and pass it to `ToSlash` as before.

As for exporting it, I'm not entirely sure. Can wait until the time it will be needed outside?

PTAL.

Reply:
here's the side-by-side - https://go.dev/play/p/UbsFcgnqvA3

Reply:
try to run that code on both Linux and Windows with `inputOS` set to "linux" and "windows". If they behave correctly, I have no strong preferences.

Reply:
Here are the two calls I'm investigating:
```
$ grep -rn "\.CheckSystemDriveAndRemoveDriveLetter("
solver/llbsolver/file/backend.go:381:   s, err := system.CheckSystemDriveAndRemoveDriveLetter(s, runtime.GOOS)
frontend/dockerfile/dockerfile2llb/convert.go:1517:             f, err := system.CheckSystemDriveAndRemoveDriveLetter(src.Path, d.platform.OS)
```

I have run both, hard-coding (swapping `linux` with `windows`) and the runs are successful for both permutations.

Results:
```
#6 [2/2] COPY test1.txt /sample/
#6 DONE 0.1s

#7 exporting to client directory
#7 copying files 19.71MB 0.1s
#7 copying files 230.88MB 0.9s done
#7 DONE 0.9s

$ ls output/sample
test1.txt
```

So, it's all good.

Reply:
For completeness, also did the reverse on Windows side (swapping `windows` with `linux`) and it's successful:

```
PS> go test -v --run=TestIntegration/TestPreserveDestDirSlash/worker=containerd 
<snip>
=== NAME  TestIntegration
    run.go:431: Skipped on windows
--- SKIP: TestIntegration (34.52s)
    --- PASS: TestIntegration/TestPreserveDestDirSlash/worker=containerd/frontend=builtin (19.11s)
    --- PASS: TestIntegration/TestPreserveDestDirSlash/worker=containerd/frontend=client (15.34s)
PASS
ok      github.com/moby/buildkit/frontend/dockerfile    34.704s
```

Reply:
buildkit can build Windows images on Linux. It uses the `inputOS` to know the target OS. So when buildkitd runs on Linux and is adding a file inside a Windows image, this code will break if the path is a Windows style path of the form `pth4 := "\\a\\b\\..\\c\\"`

Please see the last examples from here: https://go.dev/play/p/ZnAPJ2__nis

This is the reason I added the call to `cleaned := ToSlash(origPath, inputOS)` in my previous example.

Reply:
I was testing the cross-build case from Linux. Here's using the example path:

```dockerfile
FROM mcr.microsoft.com/windows/nanoserver:ltsc2022
COPY test1.txt \\a\\b\\..\\c\\
```

Build command:
```
$ docker buildx build --platform windows/amd64 `
    --builder buildkitd-dev --no-cache --tag=windows-test . `
    --progress plain `
    --output type=local,dest=./output
```

Build results:

```
#6 [2/2] COPY test1.txt \a\b\..\c\
#6 DONE 0.1s

#7 exporting to client directory
#7 copying files 21.05MB 0.1s
#7 copying files 230.88MB 0.9s done
#7 DONE 0.9s

$ ls output/
a  License.txt  ProgramData  Users  Windows
$ ls output/a/c
test1.txt
```

Reply:
Yes, this has to go through `ToSlash` for completeness. `\\a\\b\\..\\c\\`  can be a valid directly name on linux:
```sh
$ mkdir \\a\\b\\..\\c\\
$ ls -l
total 0
drwxrwxrwx 1 nandaa nandaa 4096 Oct  8 07:38 '\a\b\..\c\'
```

I was thinking that we don't presume that there could be such cases out of the wild later on. 
Right now, I just wanted to keep it simple, just a wrapper on top of `filepath.Clean()` that retains the trailing slash, especially for Windows cases both for WCOW and cross-building Windows from Linux.

Let me add the unit test cases.

Reply:
@gabriel-samfira -- I now get your point, thanks for test cases. I have brought back `inputOS` into consideration when doing `cleanPath`. PTAL at the update.

Ran the tests on both Windows and Linux:

```sh
go test -v --run=TestCheckSystemDriveAndRemoveDriveLetter ./util/system

# on windows
=== RUN   TestCheckSystemDriveAndRemoveDriveLetter
--- PASS: TestCheckSystemDriveAndRemoveDriveLetter (0.00s)
PASS
ok      github.com/moby/buildkit/util/system    0.218s

# on linux
=== RUN   TestCheckSystemDriveAndRemoveDriveLetter
--- PASS: TestCheckSystemDriveAndRemoveDriveLetter (0.00s)
PASS
ok      github.com/moby/buildkit/util/system    0.005s
```

Thanks!